lib/fauna/errors.rb

```diff
@@ -91,8 +72,97 @@ class InternalError < FaunaError; end
   # An exception thrown if FaunaDB responds with an HTTP 503.
   class UnavailableError < FaunaError; end
 
+  # :section: ErrorData
+
+  # Data for one error returned by the server.
+  class ErrorData
+    def self.from_hash(hash)
+      code = hash[:code].to_sym
+      description = hash[:description]
+      position = ErrorHelpers.map_position hash[:position]
+      if code == :'validation failed'
+        failures = hash[:failures].map(&Failure.method(:from_hash))
+        ValidationFailed.new description, position, failures
+      else
+        ErrorData.new code, description, position
+      end
+    end
+
+    ##
+    # Error code.
+    #
+    # Reference: {FaunaDB Error codes}[https://faunadb.com/documentation#errors]
+    attr_reader :code
+    # Error description.
+    attr_reader :description
+    # Position of the error in a query. May be nil.
+    attr_reader :position
+
+    def initialize(code, description, position)
+      @code = code
+      @description = description
+      @position = position
+    end
+
+    def inspect
+      "ErrorData(#{code.inspect}, #{description.inspect}, #{position.inspect})"
+    end
+  end
+
+  # An ErrorData that also stores Failure information.
+  class ValidationFailed < ErrorData
```

Comment:
I think it will be easier from the standpoint of a user trying to handle errors if failures is always on `ErrorData`, but it's just either nil (if no failures sent back) or the mapped array that's being done atm. Then you don't have to check the class type to know, and we don't have to maintain a special class for a single error code (that we don't do for any others).